internal/app/machined/pkg/controllers/cluster/endpoints_balancer.go

```diff
@@ -0,0 +1,184 @@
+	"github.com/siderolabs/talos/pkg/machinery/resources/config"
+)
+
+// EndpointsBalancerController watches v1alpha1.Config, updates discovery config.
+type EndpointsBalancerController struct{}
+
+// Name implements controller.Controller interface.
+func (ctrl *EndpointsBalancerController) Name() string {
+	return "cluster.EndpointsBalancerController"
+}
+
+// Inputs implements controller.Controller interface.
+func (ctrl *EndpointsBalancerController) Inputs() []controller.Input {
+	return []controller.Input{
+		{
+			Namespace: config.NamespaceName,
+			Type:      config.MachineTypeType,
+			ID:        pointer.To(config.MachineTypeID),
+			Kind:      controller.InputWeak,
+		},
+		safe.Input[*cluster.Affiliate](controller.InputWeak),
+		safe.Input[*config.MachineConfig](controller.InputWeak),
+		safe.Input[*cluster.Identity](controller.InputWeak),
+	}
+}
+
+// Outputs implements controller.Controller interface.
+func (ctrl *EndpointsBalancerController) Outputs() []controller.Output {
+	return []controller.Output{
+		{
+			Type: cluster.EndpointsType,
+			Kind: controller.OutputExclusive,
+		},
+	}
+}
+
+// Run implements controller.Controller interface.
+//
+//nolint:gocyclo,cyclop
+func (ctrl *EndpointsBalancerController) Run(ctx context.Context, r controller.Runtime, logger *zap.Logger) error {
+	for {
+		if _, ok := channel.RecvWithContext(ctx, r.EventCh()); !ok && ctx.Err() != nil {
+			return nil //nolint:nilerr
+		}
+
+		machineConfig, err := safe.ReaderGetByID[*config.MachineConfig](ctx, r, config.V1Alpha1ID)
+		if err != nil {
+			if !state.IsNotFoundError(err) {
+				return fmt.Errorf("error getting machine config: %w", err)
+			}
+
+			continue
+		}
+
+		machineType, err := safe.ReaderGetByID[*config.MachineType](ctx, r, config.MachineTypeID)
+		if err != nil {
+			if !state.IsNotFoundError(err) {
+				return fmt.Errorf("error getting machine type: %w", err)
+			}
+
+			continue
+		}
+
+		identity, err := safe.ReaderGetByID[*cluster.Identity](ctx, r, cluster.LocalIdentity)
+		if err != nil {
+			if !state.IsNotFoundError(err) {
+				return fmt.Errorf("error getting local identity: %w", err)
+			}
+
+			continue
+		}
+
+		affiliates, err := safe.ReaderListAll[*cluster.Affiliate](ctx, r)
+		if err != nil {
+			return fmt.Errorf("error listing affiliates: %w", err)
+		}
+
+		var endpoints []cluster.Endpoint
+
+		ce := machineConfig.Config().Cluster().Endpoint()
+		if ce != nil {
+			endpoints = append(endpoints, cluster.Endpoint{
+				Host: ce.Hostname(),
+				Port: toPort(ce.Port()),
+			})
+		}
+
+		if machineType.MachineType() == machine.TypeControlPlane {
+			endpoints = append(endpoints, cluster.Endpoint{
+				Host: "localhost",
+				Port: uint32(machineConfig.Config().Cluster().LocalAPIServerPort()),
+			})
+		}
+
+		for it := safe.IteratorFromList(affiliates); it.Next(); {
+			affiliateSpec := it.Value().TypedSpec()
+
+			if identity.TypedSpec().NodeID == affiliateSpec.NodeID {
```

Comment:
not sure if we need to skip oursleves here? does it bring any benefit?

Reply:
I think it will be kinda strange if we use both public address and localhost in the same list, no?

Reply:
I don't think it matters, if the address is local to the node, it works technically same way as `localhost`.